### PR TITLE
Use ember-remodal to simplify modal use

### PR DESCRIPTION
**Pull request by rox163**
#major#
# CHANGELOG
## Changed
- **BREAKING** Using [ember-remodal](http://sethbrasile.github.io/ember-remodal/) addon to simplify modal dialog support. Additional modal-input template options required

## Removed
- **BREAKING** Removed liquid-fire modal code. 
  - Removed the need to define/import transitions and configure the router.js file

## Upgrade notes
If you are upgrading your app to use this version, note the following - 
- Remove the modal declaration from your router.js file
- Remove any import statement from your modal component file that is importing transitions from frost-modal-input
- Remove `{{liquid-modal}}` template code if you do not have any other liquid-fire modals in your app
- Apply `remodal-bg` class to the parent container where you want a blur effect on the modal overlay.
- `frost-modal-input` root template must use block-slots and contain `ember-frost-bunsen-form` attrs - formView/formModel/formValue/onChange/onValidation. See README for details.

### CODE REVIEW

**Comment by rox163** [comment]
Seemed to have lost the styling of header+footer in the perfect scrollbar demo. Cannot figure out why. Pinged @lirwin for help.

File: package.json [hunk 1]
```diff
@@ -38,21 +40,28 @@
...
+    "ember-frost-tabs": "2.0.0",
     "ember-load-initializers": "^0.5.0",
     "ember-lodash": "0.0.7",
-    "ember-one-way-controls": "^0.6.3",
+    "ember-one-way-controls": "0.8.0",
     "ember-perfectscroll": "0.1.2",
-    "ember-prism": "^0.0.7",
+    "ember-prism": "0.0.7",
+    "ember-prop-types": "1.3.0",
+    "ember-redux": "1.3.0",
```
Comment (sglanzer) [inline package.json]: What is `redux` being used for in this addon?

File: tests/dummy/app/pods/components/frost-create-form-with-scroll/component.js [hunk 2]
```diff
@@ -60,6 +63,12 @@ export default AbstractModal.extend({
     ]
   },
 
+  clearForm () {
+    this.set('formValue', {})
+  },
+  closeModal () {
+    this.get('remodal').close('my-ps-modal')
```
Comment (sglanzer) [inline tests/dummy/app/pods/components/frost-create-form-with-scroll/component.js]: To keep the modalName aligned `'my-ps-modal'` should be provided via a variable

File: tests/dummy/app/pods/components/frost-create-form-with-scroll/template.hbs [hunk 3]
```diff
@@ -1,18 +1,22 @@
-{{#frost-modal-input title='Test title' subtitle='Subtitle'}}
-{{!#frost-modal-input titleComponent='myTitle'}}
-
-  {{#perfect-scroll
+{{#frost-modal-input
+  title='Test title'
+  subtitle='Subtitle'
+  openButton='Open Modal with Perfect Scroll'
```
Comment (sglanzer) [inline tests/dummy/app/pods/components/frost-create-form-with-scroll/template.hbs]: I'd prefer to emphasis this format

```handlebars
{{#ember-remodal as |modal|}}
    {{#modal.open}}
      {{!-- Anything you put here will open the modal on click and fire the 'onOpen' action --}}
    {{/modal.open}}
  {{/ember-remodal}}
```

Instead of the `openButton` in the attrs

Comment (sglanzer) [inline tests/dummy/app/pods/components/frost-create-form-with-scroll/template.hbs]: http://sethbrasile.github.io/ember-remodal/#/components

File: tests/dummy/app/pods/components/frost-create-form-with-scroll/template.hbs [hunk 4]
```diff
@@ -1,18 +1,22 @@
...
+  openButtonClasses='medium secondary'
+  modalName='my-ps-modal'}}
+  {{#block-slot slot 'form'}}
+    {{#perfect-scroll
```
Comment (sglanzer) [inline tests/dummy/app/pods/components/frost-create-form-with-scroll/template.hbs]: Can we move `perfect-scroll` inside of the form block-slot in the component template?

File: tests/dummy/app/pods/components/frost-create-form/component.js [hunk 5]
```diff
@@ -23,6 +26,12 @@ export default Component.extend({
     ]
   },
 
+  clearForm () {
+    this.set('userValue', {})
+  },
+  closeModal () {
+    this.get('remodal').close('my-awesome-modal')
```
Comment (sglanzer) [inline tests/dummy/app/pods/components/frost-create-form/component.js]: Same comment about promoting variables to hold the "id" so that we don't get disconnects

**Comment by sglanzer** [comment]
Can you summarize the new syntax in the README as well?

**Comment by rox163** [comment]
Yep working on the README

File: tests/dummy/app/pods/components/frost-create-form-with-scroll/template.hbs (hunk 4 above)
Comment (rox163) [inline tests/dummy/app/pods/components/frost-create-form-with-scroll/template.hbs]: I couldn't figure out how to make scroll an option because it is surrounding the form component. I could just do an if helper I suppose but that looked ugly... `if scroll true -> {{#ps}} {{form}} {{/ps}} : else -> {{form}}` ?

Comment (sglanzer) [inline tests/dummy/app/pods/components/frost-create-form-with-scroll/template.hbs]: Do we need to make it optional?  I'd prefer that the scroll is always present, but only used when we exceed the max height (I think we had that working at some point?)

Comment (rox163) [inline tests/dummy/app/pods/components/frost-create-form-with-scroll/template.hbs]: ok will look at that.

File: addon/templates/components/frost-modal-input.hbs [hunk 6]
```diff
@@ -1,11 +1,35 @@
...
+    <div class="input-header">
+      <div class="primary-title">{{title}}</div>
+      {{#if subtitle}}
+        <div class="sub-title">{{subtitle}}</div>
+      {{/if}}
+    </div>
+  {{/if}}
+  {{yield}}
+  {{#yield-slot 'form'}}
+    {{#perfect-scroll
```
Comment (rox163) [inline addon/templates/components/frost-modal-input.hbs]: tried to add this here but now the modal isnt rendering. Technically, height is 0. Form the console, Header and actions are in the source but the form component that is supposed to be yielded here isnt rendered :(

File: tests/dummy/app/pods/components/frost-create-form-with-scroll/template.hbs [hunk 7]
```diff
@@ -1,27 +0,0 @@
-{{#frost-modal-input title='Test title' subtitle='Subtitle'}}
```
Comment (rox163) [inline tests/dummy/app/pods/components/frost-create-form-with-scroll/template.hbs]: just temporarily removing this so I remember to update this at the end with the working version

File: addon/templates/components/frost-modal-input.hbs (hunk 6 above)
Comment (sglanzer) [inline addon/templates/components/frost-modal-input.hbs]: Could have sworn we had this working at some point

Comment (rox163) [inline addon/templates/components/frost-modal-input.hbs]: things may have changed with the abstract-modal component added in. Maybe Im missing something with that.

File: tests/dummy/app/pods/components/frost-create-form-with-scroll/component.js [hunk 8]
```diff
@@ -70,38 +79,24 @@ export default AbstractModal.extend({
     this.get('modalForms').setModalActive(false)
   },
```
Comment (lirwin) [inline tests/dummy/app/pods/components/frost-create-form-with-scroll/component.js]: My mistake, so if you could please add...willDestroyElement needs to call super:
  /* Ember.Component method */
  willDestroyElement () {
    this._super(...arguments)
    this.get('modalForms').setModalActive(false)
  },

File: addon/components/frost-modal-input.js [hunk 9]
```diff
@@ -1,9 +1,129 @@
...
+      Ember.$(document).off('ps-scroll-up ps-scroll-down ps-y-reach-start ps-y-reach-end')
+
+      Ember.$(window).off('resize', this.updateScrollStyles)
+
+      if (this.get('containerObserver')) {
+        this.get('containerObserver').disconnect()
+      }
+      this.set('scrollBindingsSet', false)
+    }
+  }
```
Comment (lirwin) [inline addon/components/frost-modal-input.js]: Cool...you moved the scroll bindings attach/remove to modalOpen/modalClose.  I was worried about my overlooking calling this._super() in willDestroyElement() but you have removed the need for this.  Nice refactor:)

Comment (rox163) [inline addon/components/frost-modal-input.js]: Yeah. Using ember-remodal's functions made it a bit easier

File: README.md [hunk 10]
```diff
@@ -40,7 +41,14 @@ modalForms: Ember.inject.service('modal-forms'),
 isModalActive: Ember.computed.readOnly('modalForms.isModalActive'),
 ```
 
-Change the active state of the modal in your component
+### Modal component
+Inject the [remodal](http://sethbrasile.github.io/ember-remodal/#/styling) and modal-forms services in your component
+```js
+remodal: inject.service(),
+modalForms: inject.service('modal-forms'),
```
Comment (sglanzer) [inline README.md]: Does `remodal` know when a modal is open?  E.g. could we tie to onOpen in the application controller to handle the blur and remove the `modalForms` service?

File: README.md [hunk 11]
```diff
@@ -103,68 +95,29 @@ export default AbstractModal.extend({
...
-    wheelSpeed=2
-    minScrollbarLength=20
-    maxScrollbarLength=150
-  }}
-  {{frost-bunsen-form
+{{#frost-modal-input
+  title='Test title'
+  subtitle='Subtitle'
+  modalName=modalName as |slot|}}
+  {{#block-slot slot 'openButton' as |item|}}
```
Comment (sglanzer) [inline README.md]: Do we need to yield `item` or should we just let the calling instance decide on what the launch element is?

Comment (rox163) [inline README.md]: Yeah I suppose we should support any kind of launch element, not just a button.

File: README.md [hunk 12]
```diff
@@ -103,68 +95,29 @@ export default AbstractModal.extend({
...
+    {{item.button
+      text='Open long form with scroll'
+      priority='secondary'
+      size='medium'}}
+  {{/block-slot}}
+  {{#block-slot slot 'form'}}
```
Comment (sglanzer) [inline README.md]: Let's rename this `content` since we don't interact directly with the form in the component and this makes the concept more reusable as a general modal

File: addon/styles/addon.scss [hunk 13]
```diff
@@ -12,148 +12,183 @@ $form-bkgrnd: rgba(0, 0, 0, 0);
...
-        background-color: $form-bkgrnd;
-        border: 0;
-        max-width: none;
-        padding: 0;
-      }
-    }
-  }
-}
-
 .ps-container {
```
Comment (sglanzer) [inline addon/styles/addon.scss]: I assume this should be scoped

File: addon/styles/addon.scss [hunk 14]
```diff
@@ -12,148 +12,183 @@ $form-bkgrnd: rgba(0, 0, 0, 0);
...
     margin-top: $info-bar-height + $info-bar-border-width;
+    margin-bottom: $action-bar-height + $info-bar-border-width;
   }
 }
 
-.frost-modal-input {
-  background-color: $frost-color-white;
-  border: 0;
-  box-shadow: 0 0 20 $modal-box-shadow;
+.remodal {
```
Comment (sglanzer) [inline addon/styles/addon.scss]: Similarly, I don't know if this can be scoped, but if so it would be good practice in case there are multiple types of modals (with different animations) in the system

File: addon/styles/addon.scss [hunk 15]
```diff
@@ -12,148 +12,183 @@ $form-bkgrnd: rgba(0, 0, 0, 0);
...
+                color: $frost-color-grey-5;
+                font-size: $frost-font-s;
+              }
+            }
+
+            .frost-bunsen-container {
+              margin-bottom: $action-bar-height + $form-margin-bottom;
+            }
+
+            .frost-bunsen-form {
```
Comment (sglanzer) [inline addon/styles/addon.scss]: I think the long term version of this should extract the general modal (re: comments on the `form` slot moving to `content`) into another addon and make this addon into the form-specific usage of the addon - might be worth raising an issue to track

File: README.md (hunk 10 above)
Comment (rox163) [inline README.md]: If the service is just used to blur the overlay, then I can try to just style the overlay directly? '.remodal-overlay' is available.

File: addon/templates/components/frost-modal-input.hbs [hunk 16]
```diff
@@ -1,11 +1,45 @@
...
-  <div class="input-header">
-    <div class="primary-title">{{title}}</div>
-    {{#if subtitle}}
-      <div class="sub-title">{{subtitle}}</div>
-    {{/if}}
-  </div>
-{{/if}}
-{{yield}}
+{{#ember-remodal
+  buttonClasses='frost-button'
```
Comment (sglanzer) [inline addon/templates/components/frost-modal-input.hbs]: Do we need this?

File: addon/templates/components/frost-modal-input.hbs [hunk 17]
```diff
@@ -1,11 +1,45 @@
...
+  {{yield}}
+  {{#modal.open}}
+    {{#yield-slot 'openButton'}}
+      {{yield (block-params (hash
+        button=(component 'frost-button')))}}
+    {{/yield-slot}}
+  {{/modal.open}}
+
+  {{#if titleComponent}}
+    {{component titleComponent}}
```
Comment (sglanzer) [inline addon/templates/components/frost-modal-input.hbs]: This is for legacy?  Might be a good time to make a breaking change and use the newly published version of `ember-frost-info-bar` here

Comment (sglanzer) [inline addon/templates/components/frost-modal-input.hbs]: Or just make it a `slot` and let the consumer use `ember-frost-info-bar`

File: README.md (hunk 10 above)
Comment (sglanzer) [inline README.md]: Unfortunately I don't believe the overlay is capable of a blur - the blur must be applied to the root application element in the DOM

File: addon/templates/components/frost-modal-input.hbs (hunk 17 above)
Comment (rox163) [inline addon/templates/components/frost-modal-input.hbs]: Wasnt sure if we wanted changes in the header...just basing it off usage in MCP. The only difference I see is in the padding. We have left and right padding here for the header element but info-bar only has left padding. Guessing its due to the optional action button.

File: README.md (hunk 10 above)
Comment (rox163) [inline README.md]: You are correct. I'll look into the app level blur setup.

File: addon/templates/components/frost-modal-input.hbs (hunk 17 above)
Comment (sglanzer) [inline addon/templates/components/frost-modal-input.hbs]: We can have the slot default using the `{{else}}` to what you have currently, but yield and let people use the `info-bar` otherwise.  The reason I mention this is that info-bar is getting a few tweaks to align with Marty's requests, so putting these changes behind a component makes sense so that everything remains consistent

File: addon/styles/addon.scss (hunk 15 above)
Comment (rox163) [inline addon/styles/addon.scss]: I was thinking... frost-modal-dialog [confirmation/info etc] needs this update too and maybe that can turn into the generic modal? block slots will make it easier to do.

File: README.md (hunk 10 above)
Comment (rox163) [inline README.md]: ember-remodal has built in blur support! 
`Apply the remodal-bg class to anything you want blurred in the background while a modal is open.` It tracks open state on its own! Its 3px blur by default.

File: addon/styles/addon.scss (hunk 15 above)
Comment (sglanzer) [inline addon/styles/addon.scss]: I'd prefer to keep the interface on `modal-dialog` really simple since there shouldn't be a lot of customization allowed.  The intention of that addon is to show short messages in 1 of 4 formats

File: README.md (hunk 10 above)
Comment (sglanzer) [inline README.md]: Have I mentioned that I love open source?

Comment (sglanzer) [inline README.md]: Going to go star that project in case I haven't already

File: addon/styles/addon.scss (hunk 15 above)
Comment (rox163) [inline addon/styles/addon.scss]: Ok. 
side note..not sure if you saw..There was a request to add scroll support to the dialog message content. They want to show lists of NE names/service names.

Comment (sglanzer) [inline addon/styles/addon.scss]: Hmm, let me talk to UX about this a bit, we may be able to make them align on a header/footer and scrolled content - then just make various addons to provide easy interfaces for the content section

Comment (sglanzer) [inline addon/styles/addon.scss]: Just to record the status here - this proposal is going through with UX and we should have word on this next week

**Comment by rox163** [comment]
everything updated based on our discussions, except for limiting how much the form content scrolls behind the header. Still looking into that.

File: README.md [hunk 18]
```diff
@@ -29,32 +29,60 @@ ember install ember-frost-modal-input
...
-
-### Parent Component Template
-Custom classes can be applied to the modal template, based on state of the modal
+#### Component template using ember-block-slots
+Below is an example of template.hbs
 ```handlebars
-{{liquid-modal class=(if isModalActive 'form-container' '')}}
+{{#frost-modal-input
+  modalName=modalName as |slot|}}
+  {{#block-slot slot 'openButton'}}
```
Comment (sglanzer) [inline README.md]: I think `target` is a more appropriate name for this slot

File: README.md [hunk 19]
```diff
@@ -29,32 +29,60 @@ ember install ember-frost-modal-input
...
+    {{frost-button
+      text='Open long form with scroll'
+      priority='secondary'
+      size='medium'}}
+  {{/block-slot}}
+
+  {{#block-slot slot 'header'}}
+    {{#frost-info-bar as |header|}}
```
Comment (sglanzer) [inline README.md]: `as |slot|`

File: README.md [hunk 20]
```diff
@@ -29,32 +29,60 @@ ember install ember-frost-modal-input
...
+
+  {{#block-slot slot 'form'}}
+    {{frost-bunsen-form
+      model=userModel
+      view=userView
+      onChange=(action 'formValueChanged')
+      onValidation=(action 'onValidation')}}
+  {{/block-slot}}
+
+  {{#block-slot slot 'actions'}}
```
Comment (sglanzer) [inline README.md]: `footer`?

File: README.md [hunk 21]
```diff
@@ -29,32 +29,60 @@ ember install ember-frost-modal-input
...
+      {{#block-slot header 'title'}}
+        Edit user
+      {{/block-slot}}
+      {{#block-slot header 'summary'}}
+        John Smith
+      {{/block-slot}}
+    {{/frost-info-bar}}
+  {{/block-slot}}
+
+  {{#block-slot slot 'form'}}
```
Comment (sglanzer) [inline README.md]: If we aren't tying directly to bunsen (e.g. yielding a form or taking one in the root params) then `content` is probably a better fit

File: README.md [hunk 22]
```diff
@@ -70,114 +98,32 @@ Custom classes can be applied to the modal template, based on state of the modal
...
+    this.get('remodal').close(this.get('modalName'))
   }
 })
 ```
-##### 2) Parent Component Template with ember-perfectscroll
-###### a) Wrap container element with ember-perfectscroll helper
-###### b) Footer element must have class 'actions'
-###### c) Specify options for [ember-perfectscroll](https://www.npmjs.com/package/ember-perfectscroll) helper such as wheelSpeed, minScrollbarLength, etc.
+#### Background effects
+ember-remodal provides you with a [remodal-bg](http://sethbrasile.github.io/ember-remodal/#/styling) class that you can apply to your application content, to apply blur effects to the modal overlay.
```
Comment (sglanzer) [inline README.md]: Example?

File: README.md (hunk 21 above)
Comment (rox163) [inline README.md]: yes I have 'content' in the code but didnt update the README accordingly :(

**Comment by sandersky** [comment]
It would be nice if the CHANGELOG in the description explains how the API changes after this pull request is merged. Otherwise consumers have to dig through the documentation and try to deduce what changed in order to upgrade.

File: tests/integration/components/frost-modal-input-test.js [hunk 23]
```diff
@@ -9,18 +10,48 @@ describeComponent(
...
+       Ember.run(() => {
+         application = Ember.Application.create()
+         container = application.__container__
+         application.deferReadiness()
+       initialize(container, application)
+      })
+    })
+
     it('renders', function () {
-      // Set any properties with this.set('myProperty', 'value')
```
Comment (rox163) [inline tests/integration/components/frost-modal-input-test.js]: cannot seem to get this to pass. error - `Assertion Failed: The key provided to set must be a string, you passed undefined`

File: tests/dummy/app/pods/demo/template.hbs [hunk 24]
```diff
@@ -1,14 +1,48 @@
...
+    <div class="demo">
+      {{frost-create-form
+        onConfirm=(action 'save')}}
+      {{frost-create-long-form
+        onConfirm=(action 'save')}}
+    </div>
+    <div class="snippet">
+      {{format-markdown "```handlebars
+{{#frost-modal-input
+  formView=userView
```
Comment (rox163) [inline tests/dummy/app/pods/demo/template.hbs]: moved form stuff into root template

**Comment by rox163** [comment]
@sandersky updated changelog with some upgrade notes.

File: README.md (hunk 10 above)
Comment (lirwin) [inline README.md]: Ollpoll k op mpll I'll lll look o l I'll illllllklllllol

**Comment by rox163** [comment]
bump @sglanzer

File: addon/components/frost-modal-input.js [hunk 25]
```diff
@@ -1,9 +1,140 @@
...
+  },
+
+  /**
+   * sets bindings/callbacks for perfect-scrollbar events on document element
+   * creates a window.MutationObserver for mutating events on container element
+   */
+  setScrollBindings () {
+    const $headerEl = this.get('$headerEl')
+    const $footerEl = this.get('$footerEl')
+    // const $containerEl = this.get('$containerEl')
```
Comment (juwara0) [inline addon/components/frost-modal-input.js]: Might want to remove these commented out code sections in this method. I see a few more below.

**Comment by juwara0** [comment]
Here is an issue to track increasing the code coverage on this repo: https://github.com/ciena-frost/ember-frost-modal-input/issues/25

**Comment by juwara0** [comment]
👍
